File: extension/src/sync/identity/replicaOffer.ts
export interface ReplicaOffer {
  readonly memberId: string;
  readonly admissionId: string;
  readonly identityDeviceId: string;
  readonly lanDeviceId: string;
  readonly transcriptHash: string;
  readonly endpoint: string | null;
}

const KIND = "pomo-replica-offer";

export function encodeReplicaOffer(offer: ReplicaOffer): string {
  return JSON.stringify({
    schema: 1,
    kind: KIND,
    memberId: offer.memberId,
    admissionId: offer.admissionId,
    identityDeviceId: offer.identityDeviceId,
    lanDeviceId: offer.lanDeviceId,
    transcriptHash: offer.transcriptHash,
    endpoint: offer.endpoint,
  });
}

export function decodeReplicaOffer(raw: string): ReplicaOffer {
  const value = JSON.parse(raw) as Record<string, unknown>;
  if (value.schema !== 1 || value.kind !== KIND) throw new Error("unexpected replica offer");
  return {
    memberId: hex(value.memberId, "memberId"),
    admissionId: hex(value.admissionId, "admissionId"),
    identityDeviceId: hex(value.identityDeviceId, "identityDeviceId"),
    lanDeviceId: hex(value.lanDeviceId, "lanDeviceId"),
    transcriptHash: hex(value.transcriptHash, "transcriptHash"),
    endpoint: typeof value.endpoint === "string" && value.endpoint.length > 0 ? value.endpoint : null,
  };
}

function hex(value: unknown, name: string): string {
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) throw new Error(`${name} must be 32-byte hex`);
  return value;
}

File: app/src/main/java/com/pomo/sync/identity/ReplicaOffer.kt
package com.pomo.sync.identity

import org.json.JSONObject

internal data class ReplicaOffer(
    val memberId: String,
    val admissionId: String,
    val identityDeviceId: String,
    val lanDeviceId: String,
    val transcriptHash: String,
    val endpoint: String?,
) {
    fun encode(): String =
        JSONObject()
            .put("schema", SCHEMA)
            .put("kind", KIND)
            .put("memberId", memberId)
            .put("admissionId", admissionId)
            .put("identityDeviceId", identityDeviceId)
            .put("lanDeviceId", lanDeviceId)
            .put("transcriptHash", transcriptHash)
            .put("endpoint", endpoint ?: JSONObject.NULL)
            .toString()

    companion object {
        const val SCHEMA: Int = 1
        const val KIND: String = "pomo-replica-offer"

        fun decode(raw: String): ReplicaOffer {
            val value = JSONObject(raw.trim())
            require(value.getInt("schema") == SCHEMA) { "unexpected replica offer schema" }
            require(value.getString("kind") == KIND) { "unexpected replica offer" }
            val endpoint =
                if (value.isNull("endpoint")) {
                    null
                } else {
                    value.optString("endpoint").ifBlank { null }
                }
            return ReplicaOffer(
                hex(value.getString("memberId"), "memberId"),
                hex(value.getString("admissionId"), "admissionId"),
                hex(value.getString("identityDeviceId"), "identityDeviceId"),
                hex(value.getString("lanDeviceId"), "lanDeviceId"),
                hex(value.getString("transcriptHash"), "transcriptHash"),
                endpoint,
            )
        }

        private fun hex(
            value: String,
            name: String,
        ): String {
            require(value.matches(Regex("[0-9a-f]{64}"))) { "$name must be 32-byte hex" }
            return value
        }
    }
}
